flink-libraries/flink-cep-scala/src/main/scala/org/apache/flink/cep/scala/pattern/Pattern.scala

```diff
@@ -0,0 +1,186 @@
+  }
+
+  /**
+    * Applies a subtype constraint on the current pattern operator. This means that an event has
+    * to be of the given subtype in order to be matched.
+    *
+    * @param clazz Class of the subtype
+    * @tparam S Type of the subtype
+    * @return The same pattern operator with the new subtype constraint
+    */
+  def subtype[S <: F : ClassTag](clazz: Class[S]): Pattern[T, S] = {
+    jPattern.subtype(clazz)
+    this.asInstanceOf[Pattern[T, S]]
+  }
+
+  /**
+    * Defines the maximum time interval for a matching pattern. This means that the time gap
+    * between first and the last event must not be longer than the window time.
+    *
+    * @param windowTime Time of the matching window
+    * @return The same pattern operator with the new window length
+    */
+  def within(windowTime: Time): Pattern[T, F] = {
+    jPattern.within(windowTime)
+    this
+  }
+
+  /**
+    * Appends a new pattern operator to the existing one. The new pattern operator enforces strict
+    * temporal contiguity. This means that the whole pattern only matches if an event which matches
+    * this operator directly follows the preceding matching event. Thus, there cannot be any
+    * events in between two matching events.
+    *
+    * @param name Name of the new pattern operator
+    * @return A new pattern operator which is appended to this pattern operator
+    */
+  def next(name: String): Pattern[T, T] = {
+    wrapPattern(jPattern.next(name))
+  }
+
+  /**
+    * Appends a new pattern operator to the existing one. The new pattern operator enforces
+    * non-strict temporal contiguity. This means that a matching event of this operator and the
+    * preceding matching event might be interleaved with other events which are ignored.
+    *
+    * @param name Name of the new pattern operator
+    * @return A new pattern operator which is appended to this pattern operator
+    */
+  def followedBy(name: String): FollowedByPattern[T, T] = {
+    FollowedByPattern(jPattern.followedBy(name))
+  }
+
+  /**
+    * Specifies a filter condition which has to be fulfilled by an event in order to be matched.
+    *
+    * @param filter Filter condition
+    * @return The same pattern operator where the new filter condition is set
+    */
+  def where(filter: FilterFunction[F]): Pattern[T, F] = {
+    jPattern.where(filter)
+    this
+  }
+
+  /**
+    * Specifies a filter condition which has to be fulfilled by an event in order to be matched.
+    *
+    * @param filterFun Filter condition
+    * @return The same pattern operator where the new filter condition is set
+    */
+  def where(filterFun: F => Boolean): Pattern[T, F] = {
+    val filter = new FilterFunction[F] {
+      val cleanFilter = cep.scala.cleanClosure(filterFun)
+
+      override def filter(value: F): Boolean = cleanFilter(value)
+    }
+    where(filter)
+  }
+
+  //TODO ask about java api change <?> -> <? extends T> and creating a new object vs caching object. equals/hashcode?
```

Comment:
Change to `? extends T` is good.

What do you mean with creating new object vs caching object?